### PR TITLE
Added preprocess function

### PR DESCRIPTION
**Pull request by mexxexx**
Following some discussions in issue #166 I created a preprocessing function (and an `apply_exclusion_zone` function) that is used by all (currently only 1D) algorithms. Furthermore, `core.mass` now supports time series with NaN/Inf.

### CODE REVIEW

File: stumpy/gpu_stump.py [hunk 1]
```diff
@@ -437,30 +437,32 @@ def gpu_stump(T_A, m, T_B=None, ignore_trivial=True, device_id=0):
...
-    T_B = np.asarray(T_B)
-    T_B = T_B.copy()
     if T_B.ndim != 1:  # pragma: no cover
         raise ValueError(
             f"T_B is {T_B.ndim}-dimensional and must be 1-dimensional. "
             "For multidimensional STUMP use `stumpy.mstump` or `stumpy.mstumped`"
         )
 
-    T_B[np.isinf(T_B)] = np.nan
+    core.check_dtype(T_A)
```
Comment (seanlaw) [inline stumpy/gpu_stump.py:465]: Can you please move these two lines into `core.preprocess`?

File: stumpy/stamp.py [hunk 2]
```diff
@@ -130,25 +128,22 @@ def stamp(T_A, T_B, m, ignore_trivial=False):
...
-    T_A = T_A.copy()
-    T_A[np.isinf(T_A)] = np.nan
-    core.check_dtype(T_A)
-
-    T_B = T_B.copy()
-    T_B[np.isinf(T_B)] = np.nan
     if T_B.ndim != 1:  # pragma: no cover
         raise ValueError(f"T_B is {T_B.ndim}-dimensional and must be 1-dimensional. ")
+
+    core.check_dtype(T_A)
```
Comment (seanlaw) [inline stumpy/stamp.py:141]: Please move to `core.preprocess`

File: stumpy/stomp.py [hunk 3]
```diff
@@ -107,23 +105,22 @@ def _stomp(T_A, m, T_B=None, ignore_trivial=True):
...
-
-    T_A[np.isnan(T_A)] = 0
-
     out = np.empty((l, 4), dtype=object)
 
     # Handle first subsequence, add exclusionary zone
-    if ignore_trivial:
-        P, I = stamp.mass(T_B[:m], T_A, M_T, Σ_T, 0, excl_zone)
-        PR, IR = stamp.mass(T_B[:m], T_A, M_T, Σ_T, 0, excl_zone, right=True)
+    if not np.isinf(μ_Q[0]):
```
Comment (seanlaw) [inline stumpy/stomp.py:111]: Can you please swap the order of the `if/else` statement so that it is?

```
if np.isinf(μ_Q[0]):
    .
else:
    .
```

This avoids having to perform some "mental gymnastics" to think about what `not np.isinf(μ_Q[0])` means

File: stumpy/stump.py [hunk 4]
```diff
@@ -364,29 +380,26 @@ def stump(T_A, m, T_B=None, ignore_trivial=True):
...
-    T_B = np.asarray(T_B)
-    T_B = T_B.copy()
     if T_B.ndim != 1:  # pragma: no cover
         raise ValueError(
             f"T_B is {T_B.ndim}-dimensional and must be 1-dimensional. "
             "For multidimensional STUMP use `stumpy.mstump` or `stumpy.mstumped`"
         )
-    T_B[np.isinf(T_B)] = np.nan
+
+    core.check_dtype(T_A)
```
Comment (seanlaw) [inline stumpy/stump.py:402]: Please move to `core.preprocess`

File: stumpy/stomp.py [hunk 5]
```diff
@@ -73,24 +73,22 @@ def _stomp(T_A, m, T_B=None, ignore_trivial=True):
...
+    if T_A.ndim != 1:  # pragma: no cover
+        raise ValueError(f"T_A is {T_A.ndim}-dimensional and must be 1-dimensional. ")
+
+    n = T_A.shape[0]
+
     if T_B.ndim != 1:  # pragma: no cover
         raise ValueError(f"T_B is {T_B.ndim}-dimensional and must be 1-dimensional. ")
-    T_B[np.isinf(T_B)] = np.nan
+
+    core.check_dtype(T_A)
```
Comment (seanlaw) [inline stumpy/stomp.py:91]: Please move to `core.preprocess`

File: stumpy/stumped.py [hunk 6]
```diff
@@ -83,30 +83,26 @@ def stumped(dask_client, T_A, m, T_B=None, ignore_trivial=True):
...
-    T_B = np.asarray(T_B)
     if T_B.ndim != 1:  # pragma: no cover
         raise ValueError(
             f"T_B is {T_B.ndim}-dimensional and must be 1-dimensional. "
             "For multidimensional STUMP use `stumpy.mstump` or `stumpy.mstumped`"
         )
-    T_B = T_B.copy()
-    T_B[np.isinf(T_B)] = np.nan
+
+    core.check_dtype(T_A)
```
Comment (seanlaw) [inline stumpy/stumped.py:105]: Please move to `core.preprocess`

File: tests/test_core.py [hunk 7]
```diff
@@ -32,20 +33,18 @@ def test_check_window_size():
 def naive_compute_mean_std(T, m):
```
Comment (seanlaw) [inline tests/test_core.py:33]: I have a concern that this function is no longer "naive" and is actually a copy of what we are testing against.

**Comment by mexxexx** [comment]
I honestly just forgot about scrump, there it will definitely be possible (although the exclusion zone is not necessary). For mstump I'll have to see, but it might work.

I didn't move the `check_dtype` stuff inside, because I thought it doesn't belong there.

As for the naive mean/std don't you think it's even more naive this way? Computing it in a rolling manner for each subsequence?

**Comment by seanlaw** [comment]
> I honestly just forgot about scrump, there it will definitely be possible (although the exclusion zone is not necessary). For mstump I'll have to see, but it might work.

Cool! If you can see if you can add this to `mstump`, `mstumped` then that would be great! Let's leave `scrump` out of it since I'm doing some work on it right now and we can deal with it afterward.
 
> I didn't move the check_dtype stuff inside, because I thought it doesn't belong there.

That's fair. It's more of a "check" and not really changing anything.

> As for the naive mean/std don't you think it's even more naive this way? Computing it in a rolling manner for each subsequence?

Okay, I see it now. I didn't realize when I first looked at it. I take back my "concern" 👍

**Comment by mexxexx** [comment]
How would you like to rename the function?

**Comment by seanlaw** [comment]
> How would you like to rename the function?

I don’t know 🤷‍♂️ 

So, I say let’s be pragmatic and we just go with `preprocess` for now and adjust it later if needed

**Comment by mexxexx** [comment]
Okay :smile: I included the preprocessing in MSTUMP/MSTUMPED and had to make small changes.

**Comment by seanlaw** [comment]
@mexxexx It looks like `apply_exclusion_zone` is already `njit`. Can we use this in `mstump` and `mstumped`?  I think that if we updated `_get_first_mstump_profile` and `_compute_multi_D` to use `apply_exclusion_zone` then we should be good.

**Comment by mexxexx** [comment]
I think we need a different function for `mstump`, because we have one extra dimension there.
```
D[:, zone_start : zone_stop + 1]
```
in `mstump` vs 
```
D[zone_start : zone_stop + 1]
```
in the 1D case.

**Comment by seanlaw** [comment]
According to [this](https://stackoverflow.com/questions/42916029/numpy-indexing-over-the-last-axis-when-you-dont-know-the-rank-in-advance), I think it would be fine to do:

```
zone_start = max(0, idx - excl_zone)
zone_stop = min(D.shape[0], idx + excl_zone)
D[..., zone_start : zone_stop + 1] = np.inf
```

**Comment by mexxexx** [comment]
Ah cool, I didn't know, thanks!

I'll try it out tomorrow.

**Comment by seanlaw** [comment]
@mexxexx Everything looks good. Do you think you could also rename `stamp.mass` to `stamp.mass_PI`? Then maybe we can close issue #166?

File: stumpy/stamp.py [hunk 8]
```diff
@@ -130,25 +128,22 @@ def stamp(T_A, T_B, m, ignore_trivial=False):
     T_B.shape[0]-m+1
     """
 
+    T_A, M_T, Σ_T = core.preprocess(T_A, m)
+    T_B = T_B.copy()
+    T_B[np.isinf(T_B)] = np.nan
```
Comment (seanlaw) [inline stumpy/stamp.py:133]: Is there a reason why we aren't called `core.preprocess` here?

File: tests/test_core.py [hunk 9]
```diff
@@ -171,6 +170,136 @@ def test_mass(Q, T):
...
+    left = np.linalg.norm(
+        core.z_norm(core.rolling_window(T, m), 1) - core.z_norm(Q), axis=1
+    )
+    left[np.isnan(left)] = np.inf
+
+    right = core.mass(Q, T)
+    npt.assert_almost_equal(left, right)
+
+
+def test_apply_exclusion_zone():
```
Comment (seanlaw) [inline tests/test_core.py:201]: I think it is probably important to test the "edge cases: where `index = 0` and `index = 9`

File: tests/test_core.py [hunk 10]
```diff
@@ -171,6 +170,136 @@ def test_mass(Q, T):
...
+    left = np.array([0, 1, np.inf, np.inf, np.inf, np.inf, np.inf, 7, 8, 9])
+    right = T.copy()
+    core.apply_exclusion_zone(right, index, exclusion_zone)
+
+    utils.replace_inf(left)
+    utils.replace_inf(right)
+    npt.assert_array_equal(left, right)
+
+
+def test_apply_exclusion_zone_multidimensional():
```
Comment (seanlaw) [inline tests/test_core.py:233]: Same as above: I think it is probably important to test the "edge cases: where index = 0 and index = 9

**Comment by seanlaw** [comment]
I made some mistakes trying to resolve the merge conflict. I'll fix this and also handle the renaming of `stamp.mass` as well as the extra unit tests requested above

**Comment by seanlaw** [comment]
@mexxexx Thanks again for refactoring this!
